### PR TITLE
feat: support non utf8 strings

### LINKED ISSUE #59: Loading CA cert and using for signing doesn't match original certificate
I'm trying to develop a P2P application where clients can register with a server, which will sign their certificates, so that they can later talk to each other without the involvement of the server.

I've been trying to load the CA cert using `CertificateParams::from_ca_cert_pem` and sign the client certs with `Certificate::serialize_der_with_signer`.  However the "issuer" attached to the resulting signed cert doesn't match the original cert (specifically the "subject" is 1 byte different).

I don't really know enough about TLS to figure out why this is happening, so let me know if I can provide any additional details.

My (testing) code looks like this:

```rust
    let ca_key = rcgen::KeyPair::from_der(debug_certs::CA_KEY_PK8).unwrap();
    let params = rcgen::CertificateParams::from_ca_cert_pem(
        debug_certs::CA_CERT,
        ca_key,
    )
    .unwrap();
    let ca_cert = rcgen::Certificate::from_params(params).unwrap();
    let gen_cert = rcgen::generate_simple_self_signed(vec![hostname]).unwrap();
    gen_cert.serialize_der_with_signer(&ca_cert).unwrap();
```

The issuer "subject" lines I get during certificate verification look like this (as byte arrays, note byte index 9)

Certificate "Issuer":
```
[49, 11, 48, 9, 6, 3, 85, 4, 6, 12, 2, 85, 83, 49, 17, 48, 15, 6, 3, 85, 4, 8, 12, 8, 86, 105, 114, 103, 105, 110, 105, 97, 49, 33, 48, 31, 6, 3, 85, 4, 10, 12, 24, 73, 110, 116, 101, 114, 110, 101, 116, 32, 87, 105, 100, 103, 105, 116, 115, 32, 80, 116, 121, 32, 76, 116, 100]
```
Trust Anchor "Subject":
```
[49, 11, 48, 9, 6, 3, 85, 4, 6, 19, 2, 85, 83, 49, 17, 48, 15, 6, 3, 85, 4, 8, 12, 8, 86, 105, 114, 103, 105, 110, 105, 97, 49, 33, 48, 31, 6, 3, 85, 4, 10, 12, 24, 73, 110, 116, 101, 114, 110, 101, 116, 32, 87, 105, 100, 103, 105, 116, 115, 32, 80, 116, 121, 32, 76, 116, 100]
```

[debug-game-cert.pem.txt](https://github.com/est31/rcgen/files/6644906/debug-game-cert.pem.txt)
[example_signed_cert.pem.txt](https://github.com/est31/rcgen/files/6644909/example_signed_cert.pem.txt)

### PR DESCRIPTION
**Pull request by omjadas**
fixes #59
Allows non UTF-8 values to be used for distinguished name entries

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -414,11 +427,11 @@ impl DistinguishedName {
 		removed
 	}
 	/// Inserts or updates an attribute that consists of type and name
-	pub fn push(&mut self, ty :DnType, s :impl Into<String>) {
+	pub fn push(&mut self, ty :DnType, s :DnValue) {
```
Comment (est31) [inline src/lib.rs:430]: I'd prefer to keep the old API for the push side, as 99% of use cases would probably be UTF-8 strings. This doesn't just reduce the diff of this PR but also makes updating easier for the users. Usage of the API would also remain as easy as it used to be. Can you rename this function to `push_ext` or something and say something along the lines that you should use `push` unless there is a specific need to use something other than `Utf8Strings`?

The `get` side is fine, as I think it's less used.

**Review by est31 — CHANGES_REQUESTED** [review]
Thanks for the PR! This is an API breaking change, so requires a minor semver bump for a future release. Not sure yet if I want to wait with merging or merge it right now.

File: src/lib.rs (hunk 1 above)
Comment (omjadas) [reply src/lib.rs:430]: Would it be better to have a single method that takes an `impl Into<DnValue>` and provide the following implementation?

```rust
impl<T> From<T> for DnValue
where
	T :Into<String>
{
	fn from(t :T) -> Self {
		DnValue::Utf8String(t.into())
	}
}
```

**Comment by omjadas** [comment]
@est31 I have updated the PR per your suggestions. Let me know if you would prefer `push` to be split into separate methods and I can make the change.

File: src/lib.rs (hunk 1 above)
Comment (est31) [reply src/lib.rs]: > Would it be better to have a single method that takes an impl Into<DnValue> and provide the following implementation?

That's a great idea! I'm a bit worried though that the impl is not immediately seen. A code example in the rustdoc that uses a str literal would solve that I think.

**Comment by est31** [comment]
Also I'd like to see a test case for #59. Ideally it'd not use a certificate from the wild due to copyright but one you generated yourself (maybe even using rcgen). As for the breaking change, one of the possible resolutions of #65 seems to require one already, and the ring author said they wanted to make a breaking change themselves soon.

**Comment by omjadas** [comment]
@est31 I have added the requested example and tests. Let me know if there is anything else you would like changed/added.

File: src/lib.rs [hunk 2]
```diff
@@ -414,7 +437,16 @@ impl DistinguishedName {
 		removed
 	}
 	/// Inserts or updates an attribute that consists of type and name
-	pub fn push(&mut self, ty :DnType, s :impl Into<String>) {
+	///
+    /// ```
```
Comment (est31) [inline src/lib.rs:441]: nit: formatting

File: src/lib.rs [hunk 3]
```diff
@@ -414,7 +437,16 @@ impl DistinguishedName {
...
+	/// # use rcgen::{DistinguishedName, DnType, DnValue};
+	/// let mut dn = DistinguishedName::new();
+	/// dn.push(DnType::OrganizationName, "Crab widgits SE");
+    /// dn.push(DnType::CommonName, DnValue::PrintableString("Master Cert".to_string()));
```
Comment (est31) [inline src/lib.rs:445]: nit: formatting

**Review by est31 — CHANGES_REQUESTED** [review]
OK except for nits.

File: src/lib.rs (hunk 2 above)
Comment (omjadas) [reply src/lib.rs]: fixed

File: src/lib.rs (hunk 3 above)
Comment (omjadas) [reply src/lib.rs]: fixed

**Pull request merged**